fn/async.go

```diff
@@ -0,0 +1,76 @@
+package fn
+
+import "errors"
+
+type Async[A any] interface {
+	Await() (*A, error)
+	Cancel()
+}
+
+var (
+	ContextCancelled = errors.New("context cancelled")
+	AsyncCancelled   = errors.New("async cancelled")
+)
+
+// RunAsyncWithCancel is a function that will transform a synchronous function
+// and make it a cancellable asynchronous call. This can be useful if the
+// function in the argument can block indefinitely and the calling context needs
+// a way to abandon waiting for it to complete.
+func RunAsyncWithCancel[A any](
```

Comment:
I think it's worth adding a test for this

Reply:
I'll write the test after we come to consensus on the desired API.

Reply:
i personally find a test a useful way to demonstrate how an API will be used - TDD style. 

wrote this test for testing the suggested change in the other comment.

```
func TestAsync(t *testing.T) {
	// Let test func fail internally.
	tf := newTestFunc()

	async := RunAsync[bool](tf.Run)

	tf.forceError()

	_, err := async.Await()
	require.ErrorContains(t, err, "testFunc error")

	// Let the test func succeed.
	tf = newTestFunc()

	async = RunAsync[bool](tf.Run)

	tf.complete()

	_, err = async.Await()
	require.NoError(t, err)

	// Force the func to exit due to parent context being cancelled.
	tf = newTestFunc()

	ctx := context.Background()
	ctxc, cancel := context.WithCancel(ctx)

	async = RunAsyncWithCancel[bool](ctxc, tf.Run)

	cancel()

	_, err = async.Await()
	require.ErrorContains(t, err, "context canceled")

	// Force the func to exit due to an external Cancel call.
	tf = newTestFunc()

	async = RunAsync[bool](tf.Run)

	async.Cancel()

	_, err = async.Await()
	require.ErrorContains(t, err, "context canceled")
}

func newTestFunc() *testFunc {
	return &testFunc{
		error:   make(chan struct{}),
		succeed: make(chan struct{}),
	}
}

type testFunc struct {
	error   chan struct{}
	succeed chan struct{}
}

func (f *testFunc) Run(ctx context.Context) (bool, error) {
	select {
	// Exit if parent ctx is cancelled.
	case <-ctx.Done():
		return false, ctx.Err()

	case <-f.succeed:
		return true, nil

	case <-f.error:
		return false, fmt.Errorf("testFunc error")
	}
}

func (f *testFunc) complete() {
	close(f.succeed)
}

func (f *testFunc) forceError() {
	close(f.error)
}

```